Implement join between entries for AD queries

hlecleme opened the pull request:
Allow joining LDAP entries together based on a DN attribute. The syntax is similar to the one used to give the CN of a DN attribute (e.g. `{member.cn}`) and allow to give entry attributes through an additional LDAP query (e.g. `{member.userPrincipalName}`).

The `ldap` configuration has been extended with one additional attribute `joins` to permit the user to specify the filter to be used for the query (e.g. to filter out entries not having the referenced attribute).

Thanks in advance for the review.

bersace commented:
Wow, very interesting ! Did you found somethings hard to understand in the code base ? Some remark :

- Can you write a func test or extend the func test fixtures to validate the features ?
- Can you add a Changelog entry ?
- Would you mind add a full example of join in documentation (for exemple in cookbook) ?

Did you know my naïve mock up of join feature : https://github.com/dalibo/ldap2pg/issues/128#issuecomment-347879982 ?

Thanks for this huge contribution :-)

hlecleme commented:
Thanks for the review. The ldap2pg code base is very well structured which allowed to make this change quite small.

-   Added a func test which is a variation on the default ldap2pg.yml configuration. As the AD schema is not loaded in the LDAP directory, I used the `mail` attribute instead of `sAMAccountName`. The default fixture of the LDAP database has been modified for this.
-   Changelog added
-   Added an example of configuration in the cookbook. I'm not 100% satisfied about it as it is not something that could be tried on the LDAP database configured by the default fixture, but in my opinion it is closer to a real usage.

Yes, I've seen the mock up of the join feature when investigating how to link objects together. I took some latitude as I struggled to find a link between the configuration of the mock up and the configuration found in recent versions of ldap2pg.yml. The advantages of specifying the attribute (instead of join) is to allow multiple joins from the same entry (although I do not think of a usage scenario for that) and implicit configuration of the search base. The disadvantage is that it adds one more level in the configuration of the join.

Would you prefer the implementation to follow more closely the mock up of issue #128?

bersace commented on tests/func/ldap2pg.joins.yml:
```diff
@@ -0,0 +1,143 @@
+    ORDER BY 1;
+
+
+privileges:
```
I think we can drop privileges management for this test.

hlecleme replied:
I agree with that, it does not impact what needs to be tested.

bersace commented on docs/cookbook.md:
````diff
@@ -231,6 +231,49 @@ the same result. `ldap2pg` will always execute the revoke query, thinking
+  - name: '{member.sAMAccountName}'
+```
+
+
````
Perfect :ok_hand:

bersace commented:
@hlecleme That's very good. For the format, I prefer list over dict. List item are more *embeddable* with all information. What do you think of a `using` extension like this:

``` yaml
- ldap:
    base: …
    filter: …
    join:
      using: member
      filter: …
  role:
  - name: '{sAMAccountName}'
    member: '{member.sAMAccountName}'
  - name: '{member.sAMAccountName}'
```

`join` can be a list. And `joins` is accepted as an allias as usual. You did it right.

bersace commented on ldap2pg/ldap.py:
```diff
@@ -102,24 +104,30 @@ def get_attribute(entry, attribute):
-            except KeyError:
-                raise RDNError("Unknown RDN %s" % (path[0],), raw_dn)
-
+            if path[0] in DN_COMPONENTS:
```
If I understand properly, the trick is that `member.dn` does not triggers a subquery while `member.unknownField` will. Right ? I find this elegant. :+1:

hlecleme replied:
That's right.

bersace commented on ldap2pg/manager.py:
```diff
@@ -55,6 +55,44 @@ def query_ldap(self, base, filter, attributes, scope):
+                            values_with_attrs.append((value, sub_attrs))
+                            sub_attrs_cache[(attr, value)] = sub_attrs
+                    except UserError as e:
+                        logger.warn('Ignoring %s: %s' % (value, e))
```
A detail this line should be:

``` python
logger.warning('Ignoring %s: %s', value, e)
```

`logger.warn` is deprecated by `logging`. Formatting should be done by `logging`.

hlecleme replied:
Right, sorry.

bersace commented:
Note that it's just a matter of yaml formatting. `using: member` is just a way to pass the key *in* value. This limits depth of YAML but loose the implicit unicity of each join. Still, you can map this to a dict in internal representation and thus, not changing the code).

Maybe you could accept both:

``` yaml
- ldap:
    join:
      using: member
- ldap:
    joins:
    - using: member
- ldap:
    joins:
      member:
        filter: …
...
```

I tend to accept different formats in YAML. The inner representation must always be stable (in this case, a dict). The idea is that internal representation handle all cases. But user can write only what he need to define : a single join rather than a list of one join.

Actually, lighter YAML syntax can be added later. The `join` dict is a good starting point !

bersace commented:
@hlecleme I'm ok to merge. We could improve this later (more YAML syntax, some refactoring, etc.). Can you confirm you're ready for merging ?

hlecleme commented:
@bersace Thanks again for the review. Yes, I'm ready.

bersace commented:
Bravo \o/
